File: app/src/main/java/com/zacharee1/rctdremoverforlg/misc/ProgressFragment.java
package com.zacharee1.rctdremoverforlg.misc;

import android.app.Dialog;
import android.app.DialogFragment;
import android.app.FragmentManager;
import android.content.Context;
import android.content.DialogInterface;
import android.content.res.Configuration;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.ButtonBarLayout;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Button;

import com.zacharee1.rctdremoverforlg.InstallerActivity;
import com.zacharee1.rctdremoverforlg.R;

import org.zeroturnaround.exec.ProcessExecutor;

import java.util.ArrayList;
import java.util.Arrays;

import fr.castorflex.android.circularprogressbar.CircularProgressBar;


public class ProgressFragment extends DialogFragment {
    private Runnable mPositiveRunnable;
    private Runnable mFailRunnable;

    private ProcessExecutor mExecutor = new ProcessExecutor();

    private int mTitleRes;
    private int mPositiveRes;
    private int mNegativeRes;

    private ArrayList<String> mCommands = new ArrayList<>();

    private OnOperationFinishListener mListener;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        mTitleRes = getArguments().getInt("title");
        mPositiveRes = getArguments().getInt("positive");
        mNegativeRes = getArguments().getInt("negative");
    }

    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        return new AlertDialog.Builder(getActivity())
                .setView(R.layout.terminal_output_layout)
                .setTitle(mTitleRes)
                .setPositiveButton(R.string.working, null)
                .setNegativeButton(mNegativeRes, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int i) {
                        mExecutor.stop();
                    }
                })
                .setCancelable(false)
                .create();
    }

    @Override
    public void onStart() {
        super.onStart();
        Button positive = ((AlertDialog) getDialog()).getButton(DialogInterface.BUTTON_POSITIVE);
        positive.setEnabled(false);

        ButtonBarLayout buttonBarLayout = (ButtonBarLayout) positive.getParent();
        CircularProgressBar loader = (CircularProgressBar) LayoutInflater.from(getActivity()).inflate(R.layout.loader, buttonBarLayout, false);
        loader.setId(R.id.loader);

        buttonBarLayout.addView(loader, 0);

        executeCommands();
    }

    public void show(FragmentManager manager, String tag, String... commands) {
        show(manager, tag);

        mCommands = new ArrayList<>(Arrays.asList(commands));
    }

    public void setPositiveRunnable(Runnable runnable) {
        mPositiveRunnable = runnable;
    }

    public void setFailRunnable(Runnable runnable) {
        mFailRunnable = runnable;
    }

    public void setFinishCallback(OnOperationFinishListener listener) {
        mListener = listener;
    }

    public void executeCommands() {
        new Thread(new Runnable() {
            @Override
            public void run() {
                final String result = SuUtils.suAndPrintToView(
                        mExecutor,
                        getActivity(),
                        (TerminalView) getDialog().findViewById(R.id.terminal_view),
                        mCommands
                );

                getActivity().runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        final Button positive = ((AlertDialog) getDialog()).getButton(DialogInterface.BUTTON_POSITIVE);
                        positive.setEnabled(true);

                        ((ButtonBarLayout) positive.getParent()).removeView(((ButtonBarLayout) positive.getParent()).findViewById(R.id.loader));

                        if (result.endsWith("Done!")) {
                            positive.setText(mPositiveRes);
                            positive.setOnClickListener(new View.OnClickListener() {
                                @Override
                                public void onClick(View view) {
                                    if (mPositiveRunnable != null) new Handler(Looper.getMainLooper()).post(mPositiveRunnable);
                                    dismiss();
                                }
                            });
                        } else {
                            positive.setText(R.string.retry);
                            positive.setOnClickListener(new View.OnClickListener() {
                                @Override
                                public void onClick(View view) {
                                    dismiss();
                                    if (mFailRunnable != null) new Handler(Looper.getMainLooper()).post(mFailRunnable);
                                }
                            });
                        }
                    }
                });

                if (mListener != null) mListener.onOperationFinish();
            }
        }).start();
    }

    public interface OnOperationFinishListener {
        void onOperationFinish();
    }
}

File: app/src/main/java/com/zacharee1/rctdremoverforlg/misc/ProgressFragment.kt
package com.zacharee1.rctdremoverforlg.misc

import android.annotation.SuppressLint
import android.app.Dialog
import android.content.Context
import android.content.DialogInterface
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.LayoutInflater
import android.widget.ScrollView
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.widget.ButtonBarLayout
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.FragmentManager
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.topjohnwu.superuser.CallbackList
import com.topjohnwu.superuser.Shell
import com.zacharee1.rctdremoverforlg.R
import fr.castorflex.android.circularprogressbar.CircularProgressBar
import java.util.*
import java.util.concurrent.Executors

class ProgressFragment : DialogFragment() {
    private var positiveRunnable: Runnable? = null
    private var failRunnable: Runnable? = null

    private val executor = Executors.newSingleThreadExecutor()

    private var titleRes: Int? = 0
    private var positiveRes: Int? = 0
    private var negativeRes: Int? = 0

    private var commands = ArrayList<String>()
    private var listener: OnOperationFinishListener? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        titleRes = arguments?.getInt("title")
        positiveRes = arguments?.getInt("positive")
        negativeRes = arguments?.getInt("negative")
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        isCancelable = false
        return MaterialAlertDialogBuilder(context as Context)
                .setView(R.layout.terminal_output_layout)
                .setTitle(titleRes as Int)
                .setPositiveButton(R.string.working, null)
                .setNegativeButton(negativeRes as Int) { _, _ -> executor.shutdownNow() }
                .setCancelable(false)
                .create()
    }

    override fun onStart() {
        super.onStart()
        val positive = (dialog as AlertDialog).getButton(DialogInterface.BUTTON_POSITIVE)
        positive.isEnabled = false

        val buttonBarLayout = positive.parent as ButtonBarLayout
        val loader = LayoutInflater.from(activity).inflate(R.layout.loader, buttonBarLayout, false) as CircularProgressBar
        loader.id = R.id.loader

        buttonBarLayout.addView(loader, 0)

        executeCommands()
    }

    fun show(manager: FragmentManager, tag: String, vararg commands: String) {
        show(manager, tag)

        this.commands = ArrayList(listOf(*commands))
    }

    fun setPositiveRunnable(runnable: Runnable?) {
        positiveRunnable = runnable
    }

    fun setFailRunnable(runnable: Runnable?) {
        failRunnable = runnable
    }

    fun setFinishCallback(listener: OnOperationFinishListener?) {
        this.listener = listener
    }

    @SuppressLint("CheckResult")
    fun executeCommands() {
        val outputView = dialog!!.findViewById<TerminalView>(R.id.terminal_view)
        val scroller = dialog!!.findViewById<ScrollView>(R.id.scroller)

        val callbackList = object : CallbackList<String>() {
            override fun onAddElement(e: String) {
                outputView.addText(("$e\n"))

                scroller.post { scroller.fullScroll(ScrollView.FOCUS_DOWN) }
            }
        }

        Shell.su(*commands.toTypedArray())
            .to(callbackList)
            .submit(executor) { result ->
                val positive = (dialog as AlertDialog).getButton(DialogInterface.BUTTON_POSITIVE)
                positive.isEnabled = true

                (positive.parent as ButtonBarLayout).removeView((positive.parent as ButtonBarLayout).findViewById(R.id.loader))

                if (result.isSuccess) {
                    positive.setText(positiveRes as Int)
                    positive.setOnClickListener {
                        if (positiveRunnable != null) Handler(Looper.getMainLooper()).post(positiveRunnable)
                        dismiss()
                    }
                    dialog!!.setTitle(R.string.done)
                } else {
                    positive.setText(R.string.retry)
                    positive.setOnClickListener {
                        dismiss()
                        if (failRunnable != null) Handler(Looper.getMainLooper()).post(failRunnable)
                    }
                    dialog!!.setTitle(R.string.failed)
                }

                listener?.onOperationFinish()
            }
    }

    interface OnOperationFinishListener {
        fun onOperationFinish()
    }
}